File: server/app/utils/eorzean_time.py
"""
Eorzean Time Calculation Utility
1 Eorzean hour = 175 real seconds
Multiplier: 3600/175 ≈ 20.5714
"""

from dataclasses import dataclass

EORZEA_MULTIPLIER = 3600 / 175  # ~20.5714
SECONDS_PER_DAY = 24 * 60 * 60


@dataclass
class EorzeanTime:
    hours: int
    minutes: int
    seconds: int = 0
    total_minutes: int = 0


def get_current_eorzean_time() -> EorzeanTime:
    """Get the current Eorzean Time."""
    import time
    
    # Get current Unix timestamp in seconds
    now_seconds = time.time()
    
    # Convert to Eorzean seconds
    eorzea_seconds = now_seconds * EORZEA_MULTIPLIER
    
    # Get just the time of day portion (avoid overflow by using modulo)
    eorzea_day_seconds = eorzea_seconds % SECONDS_PER_DAY
    
    hours = int(eorzea_day_seconds // 3600) % 24
    minutes = int((eorzea_day_seconds % 3600) // 60)
    seconds = int(eorzea_day_seconds % 60)
    
    return EorzeanTime(
        hours=hours,
        minutes=minutes,
        seconds=seconds,
        total_minutes=hours * 60 + minutes
    )


def get_minutes_until_et(target_hour: int) -> int:
    """Calculate minutes until a specific Eorzean hour."""
    current = get_current_eorzean_time()
    current_total = current.hours * 60 + current.minutes
    target_total = target_hour * 60
    
    diff = target_total - current_total
    if diff < 0:
        diff += 24 * 60  # Add a full ET day
    
    return diff


@dataclass
class NodeStatus:
    is_active: bool
    minutes_until_spawn: int
    status: str  # 'active', 'upcoming', 'inactive'


def get_node_status(spawn_start: int, spawn_end: int, look_ahead_hours: int = 4) -> NodeStatus:
    """
    Check if a node is currently active or spawning within the next N ET hours.
    """
    current = get_current_eorzean_time()
    current_hour = current.hours
    
    # Handle wraparound (e.g., spawn 22:00 - 02:00)
    if spawn_end >= spawn_start:
        is_active = spawn_start <= current_hour < spawn_end
    else:
        is_active = current_hour >= spawn_start or current_hour < spawn_end
    
    if is_active:
        return NodeStatus(is_active=True, minutes_until_spawn=0, status="active")
    
    # Calculate minutes until spawn
    minutes_until = get_minutes_until_et(spawn_start)
    look_ahead_minutes = look_ahead_hours * 60
    
    if minutes_until <= look_ahead_minutes:
        return NodeStatus(is_active=False, minutes_until_spawn=minutes_until, status="upcoming")
    
    return NodeStatus(is_active=False, minutes_until_spawn=minutes_until, status="inactive")


def format_et(hours: int, minutes: int = 0) -> str:
    """Format ET time for display."""
    return f"{hours:02d}:{minutes:02d}"

File: server/src/utils/eorzeanTime.js
// Eorzean Time Calculation Utility
// 1 Eorzean hour = 175 real seconds
// Multiplier: 3600/175 ≈ 20.5714

const EORZEA_MULTIPLIER = 3600 / 175;

/**
 * Get the current Eorzean Time
 * @returns {{ hours: number, minutes: number, totalMinutes: number }}
 */
export function getCurrentEorzeanTime() {
    const now = Date.now();
    const eorzeaMs = now * EORZEA_MULTIPLIER;
    const eorzeaDate = new Date(eorzeaMs);

    return {
        hours: eorzeaDate.getUTCHours(),
        minutes: eorzeaDate.getUTCMinutes(),
        totalMinutes: eorzeaDate.getUTCHours() * 60 + eorzeaDate.getUTCMinutes()
    };
}

/**
 * Calculate minutes until a specific Eorzean hour
 * @param {number} targetHour - Target ET hour (0-23)
 * @returns {number} Minutes until target hour in ET
 */
export function getMinutesUntilET(targetHour) {
    const current = getCurrentEorzeanTime();
    const currentTotalMinutes = current.hours * 60 + current.minutes;
    const targetTotalMinutes = targetHour * 60;

    let diff = targetTotalMinutes - currentTotalMinutes;
    if (diff < 0) {
        diff += 24 * 60; // Add a full ET day
    }

    return diff;
}

/**
 * Check if a node is currently active or spawning within the next N ET hours
 * @param {number} spawnStart - Node spawn start hour (ET)
 * @param {number} spawnEnd - Node spawn end hour (ET)
 * @param {number} lookAheadHours - Hours to look ahead (default 4)
 * @returns {{ isActive: boolean, minutesUntilSpawn: number, status: string }}
 */
export function getNodeStatus(spawnStart, spawnEnd, lookAheadHours = 4) {
    const current = getCurrentEorzeanTime();
    const currentHour = current.hours;
    const currentMinutes = current.minutes;

    // Handle wraparound (e.g., spawn 22:00 - 02:00)
    let isActive = false;
    if (spawnEnd >= spawnStart) {
        isActive = currentHour >= spawnStart && currentHour < spawnEnd;
    } else {
        isActive = currentHour >= spawnStart || currentHour < spawnEnd;
    }

    if (isActive) {
        return {
            isActive: true,
            minutesUntilSpawn: 0,
            status: 'active'
        };
    }

    // Calculate minutes until spawn
    const minutesUntilSpawn = getMinutesUntilET(spawnStart);
    const lookAheadMinutes = lookAheadHours * 60;

    if (minutesUntilSpawn <= lookAheadMinutes) {
        return {
            isActive: false,
            minutesUntilSpawn,
            status: 'upcoming'
        };
    }

    return {
        isActive: false,
        minutesUntilSpawn,
        status: 'inactive'
    };
}

/**
 * Format ET time for display
 * @param {number} hours
 * @param {number} minutes
 * @returns {string}
 */
export function formatET(hours, minutes = 0) {
    return `${hours.toString().padStart(2, '0')}:${minutes.toString().padStart(2, '0')}`;
}
